text/2978-stack_based_vec.md

```diff
@@ -0,0 +1,293 @@
+array_vec.push(1);
+
+// Our vector memory is now split into a 32/2016 pair of initialized and
+// uninitialized memory respectively
+assert_eq!(array_vec.len(), 1);
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`ArrayVec` is a contiguous memory block where elements can be collected, therefore, a collection by definition and even though `core::collections` doesn't exist, it is the most natural module placement.
+
+The API basically mimics most of the current `Vec` surface with some tweaks to manage capacity.
+
+Notably, these tweaked methods are fallible versions of some well-known functions like `push` that will return `Result` instead of panicking at run-time. Since the upper capacity bound is known at compile-time, the compiler is likely going to remove most of the conditional bounding checking.
+
+```rust
+// Please, bare in mind that these methods are simply suggestions. Discussions about the
+// API should probably take place elsewhere.
+
+pub struct ArrayVec<T, const N: usize> {
+    data: MaybeUninit<[T; N]>,
+    len: usize,
+}
+
+impl<T, const N: usize> ArrayVec<T, N> {
+    // Constructors
+
+    #[inline]
+    pub fn from_array(array: [T; N]) -> Self;
+
+    #[inline]
+    pub fn from_array_and_len(array: [T; N], len: usize) -> Self;
+
+    #[inline]
+    pub const fn new() -> Self;
+
+    // Methods
+
+    #[inline]
+    pub fn as_mut_ptr(&mut self) -> *mut T;
+
+    #[inline]
+    pub fn as_mut_slice(&mut self) -> &mut [T];
+
+    #[inline]
+    pub fn as_ptr(&self) -> *const T;
+
+    #[inline]
+    pub fn as_slice(&self) -> &[T];
+
+    #[inline]
+    pub const fn capacity(&self) -> usize;
+
+    pub fn clear(&mut self);
+
+    pub fn dedup(&mut self);
+
+    pub fn dedup_by<F>(&mut self, same_bucket: F)
+    where
+        F: FnMut(&mut T, &mut T) -> bool;
+
+    pub fn dedup_by_key<F, K>(&mut self, key: F)
+    where
+        F: FnMut(&mut T) -> K,
+        K: PartialEq<K>;
+
+    pub fn drain<R>(&mut self, range: R)
+    where
+        R: RangeBounds<usize>;
+
+    pub fn extend_from_cloneable_slice(&mut self, other: &[T]) -> Result<(), &[T]>
+    where
+        T: Clone;
+
+    pub fn extend_from_copyable_slice(&mut self, other: &[T]) -> Result<(), &[T]>
+    where
+        T: Copy;
+
+    pub fn insert(&mut self, _idx: usize, element: T) -> Result<(), T>;
+
+    #[inline]
+    pub const fn is_empty(&self) -> bool;
+
+    #[inline]
+    pub const fn len(&self) -> usize;
+
+    pub fn pop(&mut self) -> Option<T>;
+
+    #[inline]
+    pub fn push(&mut self, element: T) -> Result<(), T>;
+
+    pub fn remove(&mut self, idx: usize) -> Option<T>;
+
+    pub fn retain<F>(&mut self, f: F)
+    where
+        F: FnMut(&mut T) -> bool;
+
+    #[inline]
+    pub unsafe fn set_len(&mut self, len: usize);
+
+    pub fn splice<R, I>(&mut self, range: R, replace_with: I)
+    where
+        I: IntoIterator<Item = T>,
+        R: RangeBounds<usize>;
```

Comment:
I believe `splice` should return a `Splice` like vec did, same for `Drain` above.

Reply:
Yes! Thanks for pointing this out

Reply:
But the new one is still different from what `Vec` did.